### PR TITLE
Split data random

### PR DESCRIPTION
**Pull request by calvin-seamons**
So basically this branch allows the user to specify a certain percentage of C4 to train the model on. It is somewhat outlined in the template_config.yaml file.

If you have any questions let me know. I may have included some unnecessary changes but overall it should be good. Just trained a 70M model on 1% of C4 so it works at least for me

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,8 +1,8 @@
 codecarbon==2.3.4
 datasets==2.16.0
 dask==2024.2.1
-dask-expr==0.5.2
-dask-ml==2023.3.24
+dask_ml==2023.3.24
```
Comment (KimballNJardine) [inline requirements.txt:5]: Switch order to make alphabetical.

File: slurm/train_model.sh
Comment (KimballNJardine) [inline slurm/train_model.sh:1]: This should be either a new sh file or under user slurm, not replacing the existing .sh file.

File: src/train_model.py [hunk 2]
```diff
@@ -5,6 +5,7 @@
 import sys
 import torch
 import yaml
+import shutil
```
Comment (KimballNJardine) [inline src/train_model.py]: Alphabetize

File: src/train_model.py [hunk 3]
```diff
@@ -126,6 +127,51 @@ def train_model(config: Struct):
...
+
+        # Define the source and target directories
+        source_dir = Path("/grphome/grp_retnet/compute/data/datasets_tokenized/c4/train")
+        target_dir = Path("/grphome/grp_retnet/compute/data/datasets_tokenized/c4/portion_data")
+
+        # Ensure the target directory exists
+        target_dir.mkdir(parents=True, exist_ok=True)
+
+        # Calculate the number of files to copy based on the percentage
+        total_files = 1031  # Total number of parquet files in the source directory
```
Comment (KimballNJardine) [inline src/train_model.py]: This is hardcoded. Please make this work without hardcoding the number of files.

File: src/train_model.py [hunk 4]
```diff
@@ -126,6 +127,51 @@ def train_model(config: Struct):
     print(f"-log(1 / {config.vocab_size}) = " + \
         f"{-torch.log(torch.tensor(1 / config.vocab_size))}")
 
+    # Checks to see if you want to use a portion of the dataset for training
+    if config.split_dataset:
+        print(f"\nUsing {config.split_dataset_percentage}% of the dataset for training")
+
+        # Define the source and target directories
+        source_dir = Path("/grphome/grp_retnet/compute/data/datasets_tokenized/c4/train")
```
Comment (KimballNJardine) [inline src/train_model.py]: This is hardcoded. Please make this work without hardcoding the file path to the source and target directories.

File: weights/2024-01-10-16:18:17_retnet_1853472/model_args.json
Comment (KimballNJardine) [inline weights/2024-01-10-16:18:17_retnet_1853472/model_args.json:1]: Remove this file.

File: weights/2024-01-10-16:20:58_retnet_1853472/model_args.json
Comment (KimballNJardine) [inline weights/2024-01-10-16:20:58_retnet_1853472/model_args.json:1]: Remove this file.

File: weights/2024-01-10-16:38:05_retnet_1853472/model_args.json
Comment (KimballNJardine) [inline weights/2024-01-10-16:38:05_retnet_1853472/model_args.json:1]: Remove this file.

File: weights/2024-01-10-16:41:59_retnet_1853472/model_args.json
Comment (KimballNJardine) [inline weights/2024-01-10-16:41:59_retnet_1853472/model_args.json:1]: Remove this file.

File: weights/2024-01-10-18:44:47_retnet_1853472/model_args.json
Comment (KimballNJardine) [inline weights/2024-01-10-18:44:47_retnet_1853472/model_args.json:1]: Remove this file.

File: weights/2024-01-10-18:46:52_retnet_1853472/model_args.json
Comment (KimballNJardine) [inline weights/2024-01-10-18:46:52_retnet_1853472/model_args.json:1]: Remove this file.

File: weights/2024-01-12-11:39:56_retnet_1853472/model_args.json
Comment (KimballNJardine) [inline weights/2024-01-12-11:39:56_retnet_1853472/model_args.json:1]: Remove this file.

File: configs/template_config.yaml [hunk 5]
```diff
@@ -81,6 +81,13 @@ tokenized_dataset_path: "<YOUR_PATH_HERE>/data/tokenized_dataset/c4"
 # TensorBoard Path (str): Path for where TensorBoard files go, optional
 tboard_path: "<YOUR_PATH_HERE>/data/runs"
 
+# This is where you can specify what percentage of C4 data you want to use for training
```
Comment (KimballNJardine) [inline configs/template_config.yaml]: Please make this code work with datasets beyond C4 as well.

File: .gitignore [hunk 6]
```diff
@@ -35,3 +35,6 @@ data/
 **/test.json
 *.csv
 core.*
+
+# Ignore the training script
+slurm/train_model.sh
```
Comment (KimballNJardine) [inline .gitignore]: Is this correct? Maybe you mean user_slurm, or somethign like that?

**Review by KimballNJardine — CHANGES_REQUESTED** [review]
This is a quick review getting the clear issues I see. Please resolve them and then I will look more carefully at train_model.py itself.

**Comment by nprisbrey** [comment]
> This is a quick review getting the clear issues I see. Please resolve them and then I will look more carefully at train_model.py itself.

@calvin-seamons, this is a good review by @KimballNJardine. I'll take another look at it if desired after these changes have been addressed.

File: requirements.txt [hunk 7]
```diff
@@ -1,8 +1,8 @@
 codecarbon==2.3.4
 datasets==2.16.0
 dask==2024.2.1
-dask-expr==0.5.2
-dask-ml==2023.3.24
+dask_expr-1.0.3
```
Comment (nprisbrey) [inline requirements.txt]: This needs to be a `==`.

File: configs/template_config.yaml (hunk 5 above)
Comment (nprisbrey) [reply configs/template_config.yaml]: Still agree with Kimball on this one. This could be as simple as changing the comment, right @calvin-seamons?

File: configs/template_config.yaml [hunk 8]
```diff
@@ -81,6 +81,14 @@ tokenized_dataset_path: "<YOUR_PATH_HERE>/data/tokenized_dataset/c4"
...
+# IMPORTANT!!!!!
+# When set to true, you will need to go into dataset.py and edit the setup function in the DataModule class
+# Then instead of training from the train directory, you will change that to "portion_data" (About line 43)
+split_dataset: true
+split_dataset_percentage: 1
+train_model_data_path: "portion_data" # Set to "train" to use the full training directory for your model
```
Comment (nprisbrey) [inline configs/template_config.yaml]: In keeping with the current format of `template_config.yaml`, there should be an individual comment above each configuration argument with the explanation. I'd also move the inline comment for `train_model_data_path` to such an above comment as well.

File: slurm/train_model.sh
Comment (nprisbrey) [inline slurm/train_model.sh:1]: @calvin-seamons, your individual config should not override the general template!

File: src/dataset.py [hunk 9]
```diff
@@ -154,4 +155,4 @@ def pad_to_longest(self, batch):
         x_padded = torch.tensor(x_padded)
         y_true_padded = torch.tensor(y_true_padded)
 
-        return x_padded, y_true_padded
+        return x_padded, y_true_padded
```
Comment (nprisbrey) [inline src/dataset.py]: It would be nice if you'd leave EOLs at the end of a file, though I'm not sure if we're requiring it.

File: src/train_model.py [hunk 10]
```diff
@@ -2,6 +2,7 @@
 import json
 import os
 import signal
+import shutil
```
Comment (nprisbrey) [inline src/train_model.py:4]: Should be alphabetized.

**Review by nprisbrey — CHANGES_REQUESTED** [review]
@calvin-seamons -- still lacking some changes before integration, including missing a rebase/merge with main. Last I checked we're ~74 commits behind.

File: configs/template_config.yaml [hunk 11]
```diff
@@ -81,6 +81,13 @@ tokenized_dataset_path: "<YOUR_PATH_HERE>/data/tokenized_dataset/c4"
 # TensorBoard Path (str): Path for where TensorBoard files go, optional
 tboard_path: "<YOUR_PATH_HERE>/data/runs"
 
+# Split Dataset Boolean (T/F): Decide if you want to train your model on a smaller portion of C4
+split_dataset: true
+# Specify Percentage of Dataset (int): Decide what percentage of C4 you want to train on
+split_dataset_percentage: 1
```
Comment (tristancorncroissant) [inline configs/template_config.yaml]: If it's not too much of a pain to switch, can this be changed to a float? The other paramaters that represent percentages are entered as floats (e.g., `val_check_interval`, `activation_dropout`, `splits`, etc.)

File: configs/template_config.yaml [hunk 12]
```diff
@@ -81,6 +81,13 @@ tokenized_dataset_path: "<YOUR_PATH_HERE>/data/tokenized_dataset/c4"
...
+ # Train Model on Portion of Data Path (str): Set to "Train" if you want to use the full dataset, else use "portion_data"
```
Comment (tristancorncroissant) [inline configs/template_config.yaml]: This seems redundant from the `split_dataset` boolean parameter. Is it not?

File: configs/template_config.yaml (hunk 11 above)
Comment (tristancorncroissant) [reply configs/template_config.yaml]: I even think actually we could collapse both 'split_dataset' and `train_model_data_path` into `split_dataset_percentage` where default is `~`, `1`, or `1.0`, but all evaluates to the using the whole dataset. If `split_dataset_percentage != 1`, then we use the portion_data instead of train for the path and we use the decimal value as the split. The advantage being less paramters in the yaml.

File: src/train_model.py [hunk 13]
```diff
@@ -126,6 +127,51 @@ def train_model(config: Struct):
...
+        # Count the existing files in the target directory
+        existing_files_count = len(list(target_dir.glob('*.parquet')))
+
+        # Check if the existing number of files matches the expected number of files
+        if existing_files_count == files_to_copy:
+            print(f"The specified portion of the dataset ({config.split_dataset_percentage}%) is already present in {target_dir}. No changes made.")
+        else:
+            print(f"Preparing to copy {files_to_copy} files to {target_dir}.")
+
+            # Clear the target directory if the file counts do not match
```
Comment (tristancorncroissant) [inline src/train_model.py:176]: Good thinking on this block

**Comment by DrewGalbraith** [comment]
You did a lot of good work in this PR, I think it will be great to have this around to run small models.

File: configs/template_config.yaml (hunk 11 above)
Comment (KimballNJardine) [reply configs/template_config.yaml]: Agree with Drew on this.

File: src/train_model.py [hunk 14]
```diff
@@ -127,26 +127,27 @@ def train_model(config: Struct):
...
 
         # Calculate the number of files to copy based on the percentage
-        total_files = 1031  # Total number of parquet files in the source directory
-        files_to_copy = int((config.split_dataset_percentage / 100) * total_files)
+        total_files = len(list(source_dir.glob('*.parquet')))
+        files_to_copy = int((config.split_dataset) * total_files)
 
         # Count the existing files in the target directory
         existing_files_count = len(list(target_dir.glob('*.parquet')))
+        print(f"\nThere are {files_to_copy} files being copied which is actually {(files_to_copy / existing_files_count)*100}% of the training data")
```
Comment (KimballNJardine) [inline src/train_model.py]: `There are {existing_files_count} .parquet files available, of which {files_to_copy} files are being copied which is actually {(files_to_copy / existing_files_count)*100}% of the training data, instead of {config.split_dataset * 100}%` would be a better print statement

**Review by KimballNJardine — COMMENTED** [review]
I need to run the code to make sure it works. However, there are a few minor comments left that I have.

**Comment by calvin-seamons** [comment]
Testing it rn when the queue is fixed

**Comment by DrewGalbraith** [comment]
@KimballJardine @nprisbrey, where are we at on this?
